internal/service/cognitoidp/user.go

```diff
@@ -0,0 +1,584 @@
+				UserAttributes: expandUserAttributes(upd),
+			}
+			_, err := conn.AdminUpdateUserAttributes(params)
+			if err != nil {
+				if tfawserr.ErrMessageContains(err, "UserNotFoundException", "") {
+					log.Printf("[WARN] Cognito User %s is already gone", d.Id())
+					d.SetId("")
+					return nil
+				}
+				return fmt.Errorf("Error updating Cognito User Attributes: %s", err)
+			}
+		}
+		if len(del) > 0 {
+			params := &cognitoidentityprovider.AdminDeleteUserAttributesInput{
+				Username:           aws.String(d.Get("username").(string)),
+				UserPoolId:         aws.String(d.Get("user_pool_id").(string)),
+				UserAttributeNames: expandUserAttributesDelete(del),
+			}
+			_, err := conn.AdminDeleteUserAttributes(params)
+			if err != nil {
+				if tfawserr.ErrMessageContains(err, "UserNotFoundException", "") {
+					log.Printf("[WARN] Cognito User %s is already gone", d.Id())
+					d.SetId("")
+					return nil
+				}
+				return fmt.Errorf("Error updating Cognito User Attributes: %s", err)
+			}
+		}
+	}
+
+	if d.HasChange("enabled") {
+		enabled := d.Get("enabled").(bool)
+
+		if enabled {
+			enableParams := &cognitoidentityprovider.AdminEnableUserInput{
+				Username:   aws.String(d.Get("username").(string)),
+				UserPoolId: aws.String(d.Get("user_pool_id").(string)),
+			}
+			_, err := conn.AdminEnableUser(enableParams)
+			if err != nil {
+				return fmt.Errorf("Error enabling Cognito User: %s", err)
+			}
+		} else {
+			disableParams := &cognitoidentityprovider.AdminDisableUserInput{
+				Username:   aws.String(d.Get("username").(string)),
+				UserPoolId: aws.String(d.Get("user_pool_id").(string)),
+			}
+			_, err := conn.AdminDisableUser(disableParams)
+			if err != nil {
+				return fmt.Errorf("Error disabling Cognito User: %s", err)
+			}
+		}
+	}
+
+	if d.HasChange("temporary_password") {
+		password := d.Get("temporary_password").(string)
+
+		if password != "" {
+			setPasswordParams := &cognitoidentityprovider.AdminSetUserPasswordInput{
+				Username:   aws.String(d.Get("username").(string)),
+				UserPoolId: aws.String(d.Get("user_pool_id").(string)),
+				Password:   aws.String(password),
+				Permanent:  aws.Bool(false),
+			}
+
+			_, err := conn.AdminSetUserPassword(setPasswordParams)
+			if err != nil {
+				return fmt.Errorf("Error changing Cognito User's password: %s", err)
+			}
+		} else {
+			d.Set("temporary_password", nil)
+		}
+	}
+
+	if d.HasChange("password") {
+		password := d.Get("password").(string)
+
+		if password != "" {
+			setPasswordParams := &cognitoidentityprovider.AdminSetUserPasswordInput{
+				Username:   aws.String(d.Get("username").(string)),
+				UserPoolId: aws.String(d.Get("user_pool_id").(string)),
+				Password:   aws.String(password),
+				Permanent:  aws.Bool(true),
+			}
+
+			_, err := conn.AdminSetUserPassword(setPasswordParams)
+			if err != nil {
+				return fmt.Errorf("Error changing Cognito User's password: %s", err)
+			}
+		} else {
+			d.Set("password", nil)
+		}
+	}
```

Comment:
Clearing both `password` and `temporary_password` should probably call `AdminResetUserPassword` to clear the user's password

Reply:
I like your idea but it comes with additional complications: `AdminResetUserPassword` action requires users to have `email` or `phone_number` attributes set(depends on user pool configuration) when they are in the `CONFIRMED` state. And simply doesn't allow resetting password when in `FORCE_CHANGE_PASSWORD` state:

> FORCE_CHANGE_PASSWORD
```
Error: Error resetting Cognito User's password (user_pool/example): NotAuthorizedException: User password cannot be reset in the current state.
```

> CONFIRMED and no attributes
```
Error: Error resetting Cognito User's password (user_pool/example): InvalidParameterException: Cannot reset password for the user as there is no registered/verified email or phone_number
```

Are we ok with forcing tf users to set `email`/`phone_number` in case they decide to remove passwords? Also I'm not sure how to handle `User password cannot be reset in the current state` error. We somehow need to change the users's state.

Reply:
@gdavison could you provide your thoughts on this, please?

Reply:
I that case, we should leave the code as-is. It's probably worth adding a note to the documentation that clearing the `temporary_password` or `password` does not remove the password for the user